### PR TITLE
Fixes #124: Add `storage` and `selfdestruct` refunds

### LINKED ISSUE #120: Finish SSTORE operation
Consider refunds.

image

See todos here:
https://github.com/poanetwork/mana/blob/master/apps/evm/lib/evm/operation/stack_memory_storage_and_flow.ex#L114

### LINKED ISSUE #124: Pass refund general state tests
We can add these tests by updating `@passing_tests_by_group` [here](https://github.com/poanetwork/mana/blob/master/apps/blockchain/test/blockchain/state_test.exs#L9):

    @passing_tests_by_group %{
      stRefundTest: :all,
      stExample: [:add11],
      stCallCodes: [
        :callcall_00,
        :callcode_checkPC
      ]

### PR DESCRIPTION
**Pull request by masonforest**
From the yellow paper: Finally there is Ar, the refund balance,
increased through using the SSTORE instruction in order
to reset contract storage to zero from some non-zero value.
Though not immediately refunded, it is allowed to partially
offset the total execution costs.

resolves #120

### CODE REVIEW

File: apps/evm/lib/evm/interface/mock/mock_account_interface.ex [hunk 1]
```diff
@@ -110,7 +110,12 @@ defimpl EVM.Interface.AccountInterface, for: EVM.Interface.Mock.MockAccountInter
   end
 
   defp update_storage(account, key, value) do
-    put_in(account, [:storage, key], value)
+    if value == 0 do
+      pop_in(account, [:storage, key])
```
Comment (ayrat555) [inline apps/evm/lib/evm/interface/mock/mock_account_interface.ex]: ` Pipes (`|>`) can become more readable by starting with a "raw" value.`
https://github.com/rrrene/credo/blob/0b228dfbe9fc1db0787ed5e613d7df4f765c239a/lib/credo/check/refactor/pipe_chain_start.ex#L3

File: apps/blockchain/test/blockchain/state_test.exs [hunk 2]
```diff
@@ -11,6 +11,9 @@ defmodule Blockchain.StateTest do
     stCallCodes: [
       :callcall_00,
       :callcode_checkPC
+    ],
+    stRefundTest: [
+      :refund_singleSuicide
```
Comment (ayrat555) [inline apps/blockchain/test/blockchain/state_test.exs]: so currently not all tests are passing?

File: apps/evm/lib/evm/vm.ex [hunk 3]
```diff
@@ -117,6 +117,7 @@ defmodule EVM.VM do
     inputs = Operation.inputs(operation, machine_state)
 
     machine_state = MachineState.subtract_gas(machine_state, exec_env)
+    sub_state = SubState.add_refund(machine_state, sub_state, exec_env)
```
Comment (ayrat555) [inline apps/evm/lib/evm/vm.ex]: looks neat!

File: apps/blockchain/test/blockchain/state_test.exs (hunk 2 above)
Comment (masonforest) [reply apps/blockchain/test/blockchain/state_test.exs]: Nope, not yet. I [added](https://github.com/poanetwork/mana/blob/feature/refunds/apps/blockchain/test/blockchain/state_test.exs#L22-L38) and commented the failing tests and will add issues for passing the remaining tests.

File: apps/evm/test/evm/refunds_test.exs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    {:ok,
+     %{
+       account_interface: account_interface
+     }}
+  end
+
+  test "Refund: SSTORE", %{account_interface: account_interface} do
+    address = 0x0000000000000000000000000000000000000001
+
+    instructions = [
```
Comment (ayrat555) [inline apps/evm/test/evm/refunds_test.exs]: wouldn't it look better in one line?

File: apps/evm/lib/evm/refunds.ex [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  # `SSTORE` operations produce a refund when storage is set to zero from some non-zero value.
+  def refund(:sstore, [key, new_value], _machine_state, _sub_state, exec_env) do
+    old_value = ExecEnv.get_storage(exec_env, key)
+
+    if old_value != 0 && new_value == 0 do
+      @storage_refund
+    end
+  end
+
+  def refund(_opcode, _stack, _machine_state, _sub_state, _exec_env), do: nil
```
Comment (ayrat555) [inline apps/evm/lib/evm/refunds.ex]: maybe we can return zero?

File: apps/evm/lib/evm/refunds.ex [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+      iex> machine_state = %EVM.MachineState{stack: [5 , 0]}
+      iex> sub_state = %EVM.SubState{}
+      iex> EVM.Refunds.refund(machine_state, sub_state, exec_env)
+      15000
+  """
+  @spec refund(MachineState.t(), SubState.t(), ExecEnv.t()) :: t | nil
+  def refund(machine_state, sub_state, exec_env) do
+    operation = MachineCode.current_operation(machine_state, exec_env)
+    inputs = Operation.inputs(operation, machine_state)
+    refund(operation.sym, inputs, machine_state, sub_state, exec_env) || 0
```
Comment (ayrat555) [inline apps/evm/lib/evm/refunds.ex]: if we will return zero below, we can omit `|| 0`

**Comment by vyorkin** [comment]
This PR also closes #120

**Pull request merged**